stdlib/iarray.ml

```diff
@@ -56,86 +56,310 @@ type +'a t = 'a iarray
 
 (* Array operations *)
 
-external length : 'a iarray -> int = "%array_length"
-external get : 'a iarray -> int -> 'a = "%array_safe_get"
-external ( .:() ) : 'a iarray -> int -> 'a = "%array_safe_get"
-external unsafe_get : 'a iarray -> int -> 'a = "%array_unsafe_get"
+external length : local_ 'a iarray -> int = "%array_length"
+external get : ('a iarray[@local_opt]) -> int -> ('a[@local_opt]) =
+  "%array_safe_get"
+external ( .:() ) : ('a iarray[@local_opt]) -> int -> ('a[@local_opt]) =
+  "%array_safe_get"
+external unsafe_get : ('a iarray[@local_opt]) -> int -> ('a[@local_opt]) =
+  "%array_unsafe_get"
 external concat : 'a iarray list -> 'a iarray = "caml_array_concat"
+external concat_local : local_ 'a iarray list -> local_ 'a iarray =
+  "caml_array_concat_local"
 
 external append_prim : 'a iarray -> 'a iarray -> 'a iarray = "caml_array_append"
+external append_prim_local :
+  local_ 'a iarray -> local_ 'a iarray -> local_ 'a iarray =
+  "caml_array_append_local"
 external unsafe_sub : 'a iarray -> int -> int -> 'a iarray = "caml_array_sub"
+external unsafe_sub_local : local_ 'a iarray -> int -> int -> local_ 'a iarray =
+  "caml_array_sub_local"
 external unsafe_of_array : 'a array -> 'a iarray = "%array_to_iarray"
 external unsafe_to_array : 'a iarray -> 'a array = "%array_of_iarray"
 
-let init l f = unsafe_of_array (Array.init l f)
+(* Used only to reimplement [init] *)
+external unsafe_set_mutable : 'a array -> int -> 'a -> unit = "%array_unsafe_set"
+
+(* VERY UNSAFE: Can be used to violate the "no forward pointers" restriction. *)
+external make_mutable_local : int -> local_ 'a -> local_ 'a array =
+  "caml_make_local_vect"
+external unsafe_of_local_array : local_ 'a array -> local_ 'a iarray =
+  "%array_to_iarray"
+external unsafe_sub_of_array_local :
+  local_ 'a array -> int -> int -> local_ 'a iarray = "caml_array_sub_local"
+external unsafe_set_local : local_ 'a array -> int -> local_ 'a -> unit =
+  "%array_unsafe_set"
+
+(* We can't use immutable array literals here, since we don't want to require
+   the stdlib to be compiled with extensions *)
+external empty_iarray : unit -> 'a iarray =
+  "%empty_iarray"
```

Comment:
Calling a C function is more expensive, and we can't convert an empty array anymore (we used to) because we can't put the empty array on the local stack.

Reply:
Oh. That's true about the C function. Well, cross that off the list of ideas above. But there are still many other ideas there to pursue.